[flang][OpenMP] Don't allow DO CONCURRENT inside of a loop nest

tblah opened the pull request:
I don't think DO CONCURRENT fits the definition of a Canonical Loop Nest (OpenMP 6.0 section 6.4.1).
It is however explicitly allowed for the LOOP construct (6.0 section 13.8).

There's some obscure language in OpenMP 6.0 for the LOOP construct:

> If the collapsed loop is a DO CONCURRENT loop, neither the
> data-sharing attribute clauses nor the collapse clause may be specified.

From the surrounding context, I think "collapsed loop" just means the
loop that the LOOP construct applies to. So I will interpret this to
mean that DO CONCURRENT can only be used with the LOOP construct if it
does not contain the COLLAPSE clause.

This also fixes a bug where the associated clause was never cleared
after it was set.

Fixes #144178

tblah committed:
[flang][OpenMP] Don't allow DO CONCURRENT inside of a loop nest

I don't think DO CONCURRENT fits the definition of a Canonical Loop Nest
(OpenMP 6.0 section 6.4.1).

Fixes #144178

tblah committed:
Allow do concurrent in loop constructs, but not loop nests

tblah commented on flang/lib/Semantics/resolve-directives.cpp:
```diff
@@ -1952,6 +1952,12 @@ void OmpAttributeVisitor::PrivatizeAssociatedLoopIndexAndCheckLoopLevel(
+            std::get<parser::Statement<parser::NonLabelDoStmt>>(loop->t);
+        context_.Say(stmt.source,
+            "DO CONCURRENT loops cannot form part of a loop nest."_err_en_US);
+      }
```
ab3fba8 allows do concurrent as a single loop construct but not as part of a loop nest. Does that match your understanding of the standard?

kparzysz replied:
I don't think it's allowed in any worksharing-loop construct, only in the LOOP construct [6.0:424:1] "The collapsed loop may be a DO CONCURRENT loop."  No other constructs in that chapter have this annotation.

tblah committed:
Revert "Allow do concurrent in loop constructs, but not loop nests"

This reverts commit ab3fba8533682774bb752d0656e3e8c716ec5eba.

tblah committed:
Allow do concurrent with LOOP construct

There's some obscure language in OpenMP 6.0:

> If the collapsed loop is a DO CONCURRENT loop, neither the
> data-sharing attribute clauses nor the collapse clause may be specified.

From the surrounding context, I think "collapsed loop" just means the
loop that the LOOP construct applies to. So I will interpret this to
mean that DO CONCURRENT can only be used with the LOOP construct if it
does not contain the COLLAPSE clause.

This also fixes a bug where the associated clause was never cleared
after it was set.
